### PR TITLE
Refactor bindings and README

### LINKED ISSUE #440: CTRL + C won't copy on my Windows 10 machine.
I have the settings for using ctrl keys for insert and normal mode turned off. I have tried using it on WSL and windows installations of nvim. I've also tried it on VS Code and VSCodium and they all have the same results no matter what I do. Ctrl + d will highlight a word, Ctrl + V will paste and Ctrl + X will cut but all Ctrl + C does is exit insert mode. Is anyone else having the same issue?

### LINKED ISSUE #448: Visual selection: Home/End keys move cursor without updating selection
enter visual selection mode
use Home/End keys to select until end of line
expected result: selection updates the same as with ^ or $
actual result: selection stays same while only cursor moves

nvim:

NVIM v0.5.0-dev+927-g52e660e85
Build type: Debug
LuaJIT 2.1.0-beta3

vscode:
Version: 1.51.1
Commit: e5a624b788d92b8d34d1392e4c4d9789406efe8f
Date: 2020-11-10T23:31:29.624Z
Electron: 9.3.3
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Linux x64 5.8.0-31-lowlatency

### LINKED ISSUE #457: modal navigation in VSCode explorer
Love this extension! With the [VSCodeVim](https://github.com/VSCodeVim/Vim) extension, I can navigate the explorer pane (the tree view showing all files and folders) using `j` and `k`, just as in Vim. Is there a way to enable this modal motion using `j` and `k` in the tree view with this extension? Thanks!

I have been able to simulate this using the keybindings below but they require holding `ctrl` pressed. Would be nice if I didn't need to use `ctrl`.

```
{
  "key": "ctrl+j",
  "command": "list.focusDown",
  "when": "!editorFocus"
},
{
  "key": "ctrl+k",
  "command": "list.focusUp",
  "when": "!editorFocus"
}
```

### LINKED ISSUE #534: Escape does not close parameter hints or problems popup
Hey, 

Thanks for putting together this great extension. This is what got me into vim/neovim a few days ago, and I'm loving it.

I've noticed that I cannot close the popup with parameter hints or the problems popup, which is usually done using the Escape key.

For example:

Parameter hints (triggered using `gh`):
image

Problems popup (triggered using Alt + F8):
image

I assume this is because the neovim extension handles the Escape key itself and calls something along the lines of `preventDefault` on the event, which does not let VSCode handle it too.

I changed the `when` condition for the `Escape key` keybinding to:

```ts
editorTextFocus && neovim.init && !markersNavigationVisible && !parameterHintsVisible
```

(essentially, I added ` && !markersNavigationVisible && !parameterHintsVisible`)

and the Escape key started working after that.

I would happily contribute a fix to the extension if you're ok with that 🙂 I wanted to report it first and let you handle it yourself if you'd like

### PR DESCRIPTION
**Pull request by theol0403**
This is my work in progress for refactoring keybindings and readme. I hope you agree with the changes, and I will explain some of the keybindings at a later time. For now, I just want to show my progress, and see if you have any suggestions or feedback.

Would it be better to split this into two PRs? I put them in one because the readme needs to be changed to reflect the keybindings anyways.

**Todo:**

- [x] Add list/tree navigation (fixes #457)
- [x] Add file explorer actions
- [x] Make C-C binding toggleable (fixes #440)
- [x] Add navigation in peek editor (also fixes #604 and fixes #534)
- [x] Send home and end keys (fixes #448)
- [x] Refactor readme design and sentences
- [x] Convert readme large binding list into table
- [x] Document new bindings
- [x] Change ext name to "VSCode Neovim"???
- [x] Add suggested vscode settings for best keyboard navigation
- [x] Document new peek mode bindings
- [x] Add table of contents
- [ ] Provide opinionated vscode UI navigation bindings?
- [ ] Or just provide link to blog posts where people make the vscode ui very keyboard-centric
- [ ] Embrace the new wiki, move tips information over, add example navigation bindings there? Maybe after the PR.

### CODE REVIEW

**Comment by asvetliakov** [comment]
>Would it be better to split this into two PRs?

Whatever will be easier for you.

For keybindings, also can suggest these ones:
```json
  {
    "key": "ctrl+p",
    "command": "list.focusUp",
    "when": "inReferenceSearchEditor"
  },
  {
    "key": "ctrl+n",
    "command": "list.focusDown",
    "when": "inReferenceSearchEditor"
  }
```

These are for selecting prev/next reference in peek view

**Comment by Shatur95** [comment]
`=z` in Vim shows spell suggestions. How about use it for quick fix (spell extensions in VSCode also use qiuck fix for spell suggestions)?

```vim
nnoremap z= <Cmd>call VSCodeNotify('editor.action.quickFix')<CR>
```

I use the following command for multiply cursors instead of `m`:

```vim
nnoremap <C-Down> i<Cmd>call VSCodeNotify('editor.action.insertCursorBelow')<CR>
nnoremap <C-Up> i<Cmd>call VSCodeNotify('editor.action.insertCursorAbove')<CR>
```

This requires much less keystrokes. What do you think?

**Comment by Shatur95** [comment]
Also how about `gx` to open current link? This shortcut is provided by built-in netrw plugin (`:h netrw-gx`)?

```vim
nnoremap gx <Cmd>call VSCodeNotify('editor.action.openLink')<CR>
```

**Comment by David-Else** [comment]
@Shatur95 Regarding `=z` to open quickfix, it means that an integrated spellchecker could not use the correct shortcut https://github.com/asvetliakov/vscode-neovim/issues/506

`gx' already opens the current link, I just double-checked. I am using the Linux appimage on Centos 8.3.

**Comment by Shatur95** [comment]
> @Shatur95 Regarding =z to open quickfix, it means that an integrated spellchecker could not use the correct shortcut #506

Is built-in spellchecking suggestions works?

> gx already opens the current link, I just double-checked in a markdown file. I am using the Linux appimage on Centos 8.3.

Did not know that netrw handles it well :) I just have netrw disabled, so I binded this shortcut manually.

**Comment by David-Else** [comment]
> Is built-in spellchecking suggestions works?
Not really, but check out the thread... I think it could be possible in the future.

**Comment by Shatur95** [comment]
> Not really, but check out the thread... I think it could be possible in the future.

Yes, checked, thanks. I think it's okay to trigger quickfix via `z=` since future built-in spellchecking can be displayed in quickfix.

**Comment by theol0403** [comment]
Sorry for the silence... been (and will keep being) busy in the near future.

**Comment by Shatur95** [comment]
@theol0403, I think it worth to merge it as is. No one like big PR's.

**Comment by theol0403** [comment]
> `=z` in Vim shows spell suggestions. How about use it to open quick fix suggestions menu? Spell extensions in VSCode also use quick fix menu for spell suggestions.

While I do like the idea (and we can remove it if vim spellcheck gets fixed - though I think vscode spellcheck is fine), it's not very useful because you can't navigate the quickfix items with the keyboard. I think I will instead provide instructions to use the keyboard quickfix plugin, until #469 gets addressed.  

Sorry for the long delay everyone, hopefully I can get this into a mergeable state sometime soon.

**Comment by Shatur95** [comment]
> While I do like the idea (and we can remove it if vim spellcheck gets fixed - though I think vscode spellcheck is fine), it's not very useful because you can't navigate the quickfix items with the keyboard. I think I will instead provide instructions to use the keyboard quickfix plugin, until #469 gets addressed.

Sorry, I didn't mean quickfix like I did when refactoring. In this context, I mean the context menu that poups for spell suggestions. You can navigate it with keyboard.

**Comment by theol0403** [comment]
> Sorry, I didn't mean quickfix like I did when refactoring. In this context, I mean the context menu that poups for spell suggestions. You can navigate it with keyboard arrows.

Sure, even though keyboard arrows are of limited use, I don't see a reason not to add it :+1:

**Comment by theol0403** [comment]
Calling it for tonight, but end is in sight :+1:  

I took a lot of liberty in this PR by changing whatever I saw fit, please feel free to veto anything you aren't comfortable with.

Todo in OP has been updated.

File: package.json
Comment (trkoch) [inline package.json:741]: Would this break native scroll navigation when pressing Home/End?

File: vim/vscode-code-actions.vim
Comment (trkoch) [inline vim/vscode-code-actions.vim:98]: Not sure this is something of universal use. Users can remap this easily in their local installation.

File: package.json
Comment (trkoch) [inline package.json:397]: So this is for navigating lists with j/k?

**Review by trkoch — COMMENTED** [review]
Maybe you could split this PR and get the non-controversial parts merged?

File: package.json
Comment (theol0403) [reply package.json:397]: Yes, it is. Taken from VSCodeVim.

File: vim/vscode-code-actions.vim
Comment (theol0403) [reply vim/vscode-code-actions.vim:98]: While I do agree, I think that it does not hurt to have some default functionality for a common keybinding, rather than it not working at all.

File: package.json
Comment (theol0403) [reply package.json:741]: Not sure what you mean. In vscode, home/end moves to the beginning and end of the line - this is what neovim also does.

File: package.json
Comment (theol0403) [reply package.json]: I'm open to discussion, but I guess I was trying to make the "branding" more consistent. "VSCode Neovim - VSCode Neovim Integration". I changed it in my last few commits, let me know if you still prefer the old version.

**Comment by theol0403** [comment]
> Maybe you could split this PR and get the non-controversial parts merged?

I feel bad for taking forever to complete this PR, but I think everything should be done at once - the keybindings are coupled with the readme, and I don't want to work on the readme in two passes (write docs for the new bindings in the old format, and then convert to the new format). And then, since the whole readme is being refactored, it would also be a good time to change the branding and whatever.

I will be ready to submit this PR for review after I complete the new peek mode bindings and adding a table of contents. Probably I can get that done in the next couple days.

After this is merged, I'll also look into moving a lot of the information into the wiki, and possibly adding more stuff in terms of opinionated mappings - the readme is already getting very big.

**Comment by theol0403** [comment]
I have merged #604 into my PR, I hope they still get recognition for it. I did it because I needed to modify it to also close the peek editor.

**Comment by asvetliakov** [comment]
I'm eager to merge it once it done. Awesome!

**Comment by theol0403** [comment]
Well, I think everything I plan to do is complete 🎉!

However, give me a day or two to just go over it and make sure everything is perfect. I think the only thing I might change is the order of some bindings in the readme, and possibly squash some meaningless commits.

If you want to start the review, however, feel free!

**Comment by theol0403** [comment]
I have made some more improvements, and rebased/squashed some of my commits to make it easier to read and know what changed. 

I think this is ready to be merged! 🎉 

Of course, please make sure you agree with all the decisions I made before merging 👍

**Review by asvetliakov — APPROVED** [review]

**Comment by theol0403** [comment]
Sorry for all the force pushes, I just keep finding little things I want to change that don't deserve a new commit 😅.

Do you want me to press the merge button, or do you want to do it?

**Comment by asvetliakov** [comment]
Go 🚗 😄

**Comment by theol0403** [comment]
Last question, do you want to squash, or keep commits as-is? The latter might be better, to better isolate the individual changes, but there is also a bit of commit mess.

**Comment by asvetliakov** [comment]
it's not very important. squashing helps when you're doing some kind of deploy-per-commit, but it's not relevant to this repo. so whatever you feel right. I'd squash it

**Comment by theol0403** [comment]
Finally!! Hopefully people will like the new changes 😄 

Thanks again for your awesome work on this plugin :heart:! I hope to see it grow in the future.

Just a small thing, can you capitalize the i in "VSCode Neovim integration" in the github description to stay consistent?

**Comment by asvetliakov** [comment]
awesome work! thank you

>Just a small thing, can you capitalize the i in "VSCode Neovim integration" in the github description to stay consistent?
done

**Pull request merged**
